### PR TITLE
[FIRRTL] Keep InstancePathCache updated

### PR DESCRIPTION
**Pull request by prithayan**
Add helper to replace an old `InstanceOp` with a new `InstanceOp` to keep the InstancePathCache updated.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLInstanceGraph.cpp [hunk 1]
```diff
@@ -71,3 +71,40 @@ InstancePath InstancePathCache::appendInstance(InstancePath path,
...
+void InstancePathCache::replaceInstance(InstanceOp oldOp, InstanceOp newOp) {
+
+  instanceGraph.replaceInstance(oldOp, newOp);
+
+  // Iterate over all the paths, and search for the old InstanceOp. If found,
+  // then replace it with the new InstanceOp, and create a new copy of the paths
+  // and update the cache.
+  auto instanceExists = [&](const ArrayRef<InstancePath> &paths) -> bool {
+    return llvm::any_of(paths, [&](InstancePath p) {
+      return llvm::any_of(p, [&](InstanceOp inst) { return inst == oldOp; });
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLInstanceGraph.cpp:84]:
```suggestion
      return llvm::is_contained(p, oldOp);
```

File: lib/Dialect/FIRRTL/FIRRTLInstanceGraph.cpp [hunk 2]
```diff
@@ -71,3 +71,40 @@ InstancePath InstancePathCache::appendInstance(InstancePath path,
...
+    SmallVector<InstancePath, 8> updatedPaths;
+    for (auto path : iter.getSecond()) {
+      const auto *iter = llvm::find(path, oldOp);
+      if (iter == path.end()) {
+        // path does not contain the oldOp, just copy it as is.
+        updatedPaths.push_back(path);
+        continue;
+      }
+      auto *newPath = allocator.Allocate<InstanceOp>(path.size());
+      std::copy(path.begin(), path.end(), newPath);
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLInstanceGraph.cpp:100]:
```suggestion
      llvm::copy(path, newPath);
```

File: lib/Dialect/FIRRTL/FIRRTLInstanceGraph.cpp [hunk 3]
```diff
@@ -71,3 +71,40 @@ InstancePath InstancePathCache::appendInstance(InstancePath path,
...
+      newPath[iter - path.begin()] = newOp;
+      updatedPaths.push_back(InstancePath(newPath, path.size()));
+    }
+    // Move the list of paths into the bump allocator for later quick
+    // retrieval.
+    auto *paths = allocator.Allocate<InstancePath>(updatedPaths.size());
+    std::copy(updatedPaths.begin(), updatedPaths.end(), paths);
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLInstanceGraph.cpp:107]:
```suggestion
    std::copy(updatedPaths, paths);
```

File: include/circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h [hunk 4]
```diff
@@ -75,6 +75,12 @@ struct InstancePathCache {
       : instanceGraph(instanceGraph) {}
   ArrayRef<InstancePath> getAbsolutePaths(Operation *op);
 
+  /// Clear the cache.
+  void invalidate() { absolutePathsCache.clear(); }
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h:79]: Should this also reset the allocator?

**Comment by prithayan** [comment]
Thanks @dtzSiFive for the review, updated to reset the allocator and merged it.

**Pull request merged**